core/embed/rust/src/ui/shape/render.rs

```diff
@@ -0,0 +1,252 @@
+            canvas.fill_background(color);
+        }
+
+        // TODO: consider storing original canvas.viewport
+        //       and restoring it by drop() function
+
+        Self { canvas, cache }
+    }
+}
+
+impl<'a, 'alloc, C> Renderer<'alloc> for DirectRenderer<'a, 'alloc, C>
+where
+    C: Canvas,
+{
+    fn viewport(&self) -> Viewport {
+        self.canvas.viewport()
+    }
+
+    fn set_viewport(&mut self, viewport: Viewport) {
+        self.canvas.set_viewport(viewport);
+    }
+
+    fn render_shape<S>(&mut self, mut shape: S)
+    where
+        S: Shape<'alloc> + ShapeClone<'alloc>,
+    {
+        if self.canvas.viewport().contains(shape.bounds(self.cache)) {
+            shape.draw(self.canvas, self.cache);
+            shape.cleanup(self.cache);
+        }
+    }
+}
+
+// ==========================================================================
+// struct ProgressiveRenderer
+// ==========================================================================
+
+struct ShapeHolder<'a> {
+    shape: &'a mut dyn Shape<'a>,
+    viewport: Viewport,
+}
+
+/// A more advanced Renderer implementation that supports deferred rendering.
+pub struct ProgressiveRenderer<'a, 'alloc, T, C>
+where
+    T: LocalAllocLeakExt<'alloc>,
+    C: BasicCanvas,
+{
+    /// Target canvas
+    canvas: &'a mut C,
+    /// Bump for cloning shapes
+    bump: &'alloc T,
+    /// List of rendered shapes
+    shapes: FixedVec<'alloc, ShapeHolder<'alloc>>,
+    /// Current viewport
+    viewport: Viewport,
+    // Default background color
+    bg_color: Option<Color>,
+    /// Drawing cache (decompression context, scratch-pad memory)
+    cache: &'a DrawingCache<'alloc>,
+}
+
+impl<'a, 'alloc, T, C> ProgressiveRenderer<'a, 'alloc, T, C>
+where
+    T: LocalAllocLeakExt<'alloc>,
+    C: BasicCanvas,
+{
+    /// Creates a new ProgressiveRenderer instance
+    pub fn new(
+        canvas: &'a mut C,
+        bg_color: Option<Color>,
+        cache: &'a DrawingCache<'alloc>,
+        bump: &'alloc T,
+        max_shapes: usize,
+    ) -> Self {
+        let viewport = canvas.viewport();
+        Self {
+            canvas,
+            bump,
+            shapes: unwrap!(bump.fixed_vec(max_shapes), "No shape memory"),
+            viewport,
+            bg_color,
+            cache,
+        }
+    }
+
+    /// Renders stored shapes onto the specified canvas
+    pub fn render(&mut self, lines: usize) {
+        let canvas_clip = self.canvas.viewport().clip;
+
+        if canvas_clip.is_empty() {
+            return;
+        }
+
+        let buff = &mut unwrap!(self.cache.render_buff(), "No render buffer");
+
+        let mut slice = unwrap!(
+            Rgb565Canvas::new(
+                Offset::new(canvas_clip.width(), lines as i16),
+                None,
+                Some(1),
+                &mut buff[..],
+            ),
+            "No render memory"
+        );
+
+        for y in (canvas_clip.y0..canvas_clip.y1).step_by(lines) {
```

Comment:
for instance, this sure looks like it relies on `slice` being `lines` tall
perhaps this is missing something like `let lines = slice.size().y` ?

Reply:
good catch, fixed, see 75cceb3cfd117f49dcb93960b2e25a5e837abff5